src/lem/circuit.rs

```diff
@@ -708,13 +734,85 @@ impl Func {
                         unhash_helper!(preimg, img, SlotType::Hash4);
                     }
                     Op::Null(tgt, tag) => {
+                        let tag = g.global_allocator.get_or_alloc_const(cs, tag.to_field())?;
+                        let zero = g.global_allocator.get_or_alloc_const(cs, F::ZERO)?;
+                        let allocated_ptr = AllocatedPtr::from_parts(tag, zero);
+                        bound_allocations.insert(tgt.clone(), allocated_ptr);
+                    }
+                    Op::Lit(tgt, lit) => {
+                        let lit_ptr = lit.to_ptr(g.store);
+                        let lit_tag = lit_ptr.tag().to_field();
+                        let lit_hash = g.store.hash_ptr(&lit_ptr)?.hash;
+                        let allocated_tag = g.global_allocator.get_or_alloc_const(cs, lit_tag)?;
+                        let allocated_hash = g.global_allocator.get_or_alloc_const(cs, lit_hash)?;
+                        let allocated_ptr = AllocatedPtr::from_parts(allocated_tag, allocated_hash);
+                        bound_allocations.insert(tgt.clone(), allocated_ptr);
+                    }
+                    Op::Cast(tgt, tag, src) => {
+                        let src = bound_allocations.get(src)?;
+                        let tag = g.global_allocator.get_or_alloc_const(cs, tag.to_field())?;
+                        let allocated_ptr = AllocatedPtr::from_parts(tag, src.hash().clone());
+                        bound_allocations.insert(tgt.clone(), allocated_ptr);
```

Comment:
Okay, but this can only be used select few places. Many supported conversions are not strict 'casts' like what you've written here. We've discussed elsewhere, so I assume this is accounted for.

Reply:
Yes, we can't cast Nums to U64 just by changing tags, but LEM's cast is lower level, so it will allow you to do "wrong" things like that. I think that LEM should ultimately not understand tags, and it's up to the LEM programmer to give a meaning to tags. I eventually plan to make LEM generic by defining some traits that define what kinds of constants are expressible and how it is converted to circuit (by providing is_field) and so on